theseus/embodied/robot/FK.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+from typing import List, Optional
+
+from theseus.geometry.functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+# TODO: Add support for joints with DOF>1
+
+
+def ForwardKinematicFactory(robot: Robot, link_names: Optional[List[str]] = None):
+    links: List[Link] = []
+
+    if link_names is None:
+        links = robot.links
+    else:
+        links = [robot.link_map[name] for name in link_names]
+
+    ancestors = []
+
+    for link in links:
+        ancestors += [anc.id for anc in link.ancestors]
+
+    link_ids = sorted(list(set(ancestors)))
+
+    def ForwardKinematicsImpl(angles: torch.Tensor):
+        if angles.ndim != 2 or angles.shape[1] != robot.dof:
+            raise ValueError(
+                f"Joint angles for {robot.name} should be {robot.dof}-D vectors"
+            )
+
+        poses: List[Optional[torch.Tensor]] = [None] * robot.num_links
+        poses[0] = angles.new_zeros(angles.shape[0], 3, 4)
+        poses[0][:, 0, 0] = 1
+        poses[0][:, 1, 1] = 1
+        poses[0][:, 2, 2] = 1
+
+        for id in link_ids[1:]:
+            curr: Link = robot.links[id]
+            joint: Joint = robot.links[id].parent
+            prev: Link = joint.parent
+            relative_pose = (
+                joint.relative_pose(angles[:, joint.id])
+                if joint.id < robot.dof
+                else joint.relative_pose()
```

Comment:
I'm worried that if joint ID assignment is not closely adhering to some standard this might break. Is there a more systematic way to map the vector joint states to individual joints? 
One possibility I'm thinking is to decouple the topological relationship of links and joints with their IDs. The IDs are used to order state and action vectors, while FK and other kinematic / dynamic computations directly operate on the tree data structure.

Thus, there also won't be the need to "process fixed joints after non-fixed ones to make their IDs larger value"

Reply:
Yeah, I do notice that some libraries decouple link and joint ids. The most immediate benefit to associate link and joint ids improves the computational efficiency. For example, we could get the parent joint id of a link. As far as I know, lots of fast kinematics and dynamics libraries construct the kinematic tree in this way.

Reply:
I'm not sure I fully understand, do you mean that getting the some ancestor joint id is now constant time (joint.id - X) to traverse X steps, instead of needing to backtrack X times? But that only applies to kinematic chains and would break with trees?
Let's discuss in more detail during the meeting today.